### PR TITLE
[BC Break] Cleanup dataProvider code

### PR DESCRIPTION
**Pull request by fzaninotto**
We no longer need to rely on Redux for the data fetching.

- [x] Remove data provider actions
- [x] Remove data provider reducers
- [x] Remove data provider optimistic / pessimistic query logic
- [x] Remove base data provider hooks (`useQuery`, `useQueryWithStore`, `useMutation`)
- [x] Remove data provider base components (`<Query>`, `<Mutation>`)
- [x] Fix tests
- [x] Write doc
- [x] Write upgrade guide

### CODE REVIEW

File: packages/ra-core/src/dataProvider/index.ts [hunk 1]
```diff
@@ -45,7 +45,5 @@ export {
     useCreate,
     useDelete,
     useDeleteMany,
-    useRefreshWhenVisible,
     withDataProvider,
-    useIsAutomaticRefreshEnabled,
```
Comment (fzaninotto) [inline packages/ra-core/src/dataProvider/index.ts:50]: Note for self: uin the EditController, pass an option to the useGetOne call to disable auto refresh on the record data, as it creates unexpected effects on forms.

**Comment by fzaninotto** [comment]
The PR is now Ready for Review

File: UPGRADE.md [hunk 2]
```diff
@@ -108,6 +108,193 @@ This should be mostly transparent for you unless:
...
+```
+
+If you're calling a custom dataProvider method, then you can use react-query's `useQuery`or `useMutation` instead:
+
+```diff
+-import { useMutation } from 'react-admin';
++import { useDataProvider } from 'react-admin';
++import { useMutation } from 'react-query';
+const BanUserButton = ({ userId }) => {
+-   const { loaded, error, data } = useMutation({
```
Comment (WiXSL) [inline UPGRADE.md:148]:
```suggestion
-   const [mutate, { loading, error, data }] = useMutation({
```

File: UPGRADE.md [hunk 3]
```diff
@@ -108,6 +108,193 @@ This should be mostly transparent for you unless:
...
+-       type: 'banUser',
+-       payload: userId
```
Comment (WiXSL) [inline UPGRADE.md:150]:
```suggestion
-       payload: { id: userId }
```

File: UPGRADE.md [hunk 4]
```diff
@@ -108,6 +108,193 @@ This should be mostly transparent for you unless:
...
+-   });
++   const dataProvider = useDataProvider();
++   const { mutate, isLoading } = useMutation(
++       ['banUser', userId],
++       () => dataProvider.banUser(userId)
++   );
+    return <Button label="Ban" onClick={() => mutate()} disabled={isLoading} />;
```
Comment (WiXSL) [inline UPGRADE.md:157]:
```suggestion
-   return <Button label="Ban" onClick={() => mutate()} disabled={loading} />;
+   return <Button label="Ban" onClick={() => mutate()} disabled={isLoading} />;
```

File: UPGRADE.md [hunk 5]
```diff
@@ -108,6 +108,193 @@ This should be mostly transparent for you unless:
...
+## `useDataProvider` No Longer Accepts Side Effects
+
+`useDataProvider` returns a wrapper around the application `dataProvider` instance. In previous react-admin versions, the wrapper methods used to accept an `options` object, allowing to pass `onSuccess` and `onFailure` callbacks. This is no longer the case - the wrapper returns an object with the same method signatures as the original `dataProvider`.
+
+If you need to call the `dataProvider` and apply side effects, use react-query's `useQuery` or `useMutation` hooks instead.
+
+```diff
+import { useDataProvider } from 'react-admin';
++import { useMutation } from 'react-query';
+
```
Comment (WiXSL) [inline UPGRADE.md:200]:
```suggestion
-import { useState } from 'react';

```

File: UPGRADE.md [hunk 6]
```diff
@@ -108,6 +108,193 @@ This should be mostly transparent for you unless:
...
+const BanUserButton = ({ userId }) => {
+    const dataProvider = useDataProvider();
++   const { mutate, isLoading } = useMutation();
```
Comment (WiXSL) [inline UPGRADE.md:203]:
```suggestion
+   const { mutate, isLoading } = useMutation();
-   const [loading, setLoading] = useState(false);

```

File: UPGRADE.md [hunk 7]
```diff
@@ -108,6 +108,193 @@ This should be mostly transparent for you unless:
...
+    const handleClick = () => {
+-       dataProvider.banUser(userId, { onSuccess: () => console.log('User banned') });
```
Comment (WiXSL) [inline UPGRADE.md:205]:
```suggestion
-       setLoading(true);
-       dataProvider.banUser(userId, {
-           onSuccess: () => {
-               setLoading(false);
-               console.log('User banned');
-           },
-       });
```

File: UPGRADE.md [hunk 8]
```diff
@@ -108,6 +108,193 @@ This should be mostly transparent for you unless:
...
++       mutate(
++           ['banUser', userId],
++           () => dataProvider.banUser(userId),
++           { onSuccess: () => console.log('User banned') }
++       );
+    }
+    return <Button label="Ban" onClick={handleClick} disabled={isLoading} />;
```
Comment (WiXSL) [inline UPGRADE.md:212]:
```suggestion
-   return <Button label="Ban" onClick={handleClick} disabled={loading} />;
+   return <Button label="Ban" onClick={handleClick} disabled={isLoading} />;
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: UPGRADE.md (hunk 5 above)
Comment (WiXSL) [reply UPGRADE.md:200]: Added loading state to show how to change it to disable the button below

**Review by WiXSL — CHANGES_REQUESTED** [review]
Remove / Update references to removed actions:

https://github.com/marmelab/react-admin/blob/0de1637a5605684a8977a0d98b8977ceff645e62/packages/ra-core/src/dataProvider/useDataProvider.ts#L80

File: packages/ra-core/src/reducer/admin/resource/list/validity.ts [hunk 9]
```diff
@@ -1,5 +1,5 @@
 import { Reducer } from 'redux';
-import { FETCH_END, REFRESH_VIEW } from '../../../../actions';
```
Comment (WiXSL) [inline packages/ra-core/src/reducer/admin/resource/list/validity.ts:2]: Resolve `FETCH_END` usage

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: packages/ra-core/src/reducer/admin/resource/list/validity.ts [hunk 10]
```diff
@@ -1,5 +1,5 @@
...
+import { FETCH_END } from '../../../../actions';
```
Comment (WiXSL) [inline packages/ra-core/src/reducer/admin/resource/list/validity.ts:2]: Resolve FETCH_END usage

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Review by WiXSL — CHANGES_REQUESTED** [review]
Update comment as mutation callbacks can no longer be used as event handlers:

```diff
- return <button disabled={isLoading} onClick={updateMany}>Reset views</button>;
+ return <button disabled={isLoading} onClick={() => updateMany()}>Reset views</button>;
```
https://github.com/marmelab/react-admin/blob/0de1637a5605684a8977a0d98b8977ceff645e62/packages/ra-core/src/dataProvider/useUpdateMany.ts#L65

**Review by WiXSL — CHANGES_REQUESTED** [review]
Typo

```diff
- For instance, here is how to override the side eggects for the `getOne` query in a `<Show>` component: 
+ For instance, here is how to override the side effects for the `getOne` query in a `<Show>` component: 
```

https://github.com/marmelab/react-admin/blob/0de1637a5605684a8977a0d98b8977ceff645e62/UPGRADE.md?plain=1#L543

**Review by WiXSL — COMMENTED** [review]
It's still valid to pass an `onSuccess` side effect like in this comment ?

```
* May include side effects to be executed upon success or failure, e.g. { onSuccess: { refresh: true } }

**Review by WiXSL — CHANGES_REQUESTED** [review]
Remove `useVersion` hook from `Reference.md` line 212
```diff
- * `useVersion` 
```

**Review by WiXSL — CHANGES_REQUESTED** [review]
Edit's `successMessage` is still reference in some files and should be removed:

- `docs/CreateEdit.md`
- `packages/ra-core/src/controller/edit/useEditContext.tsx`
- `packages/ra-ui-materialui/src/detail/Edit.tsx`
- `packages/ra-ui-materialui/src/detail/EditView.tsx`

File: UPGRADE.md (hunk 3 above)
Comment (fzaninotto) [reply UPGRADE.md:150]: Not necessarily, and nto the purpose of this snippet.

**Comment by fzaninotto** [comment]
Review applied

**Review by djhi — APPROVED** [review]
There an eslint warning left but otherwise, this is great!

**Pull request merged**